src/clojure/nrepl/middleware/dynamic_loader.clj

```diff
@@ -0,0 +1,108 @@
+(ns nrepl.middleware.dynamic-loader
+  "Support the ability to interactively update the middleware of the *running*
+  nREPL server. This can be used by tools to configure an existing instance of
+  an environment after connection.
+
+  When combined with the sideloader, this could be used to inject middlewares
+  that are unknown to the server prior to connection."
+  {:author "Shen Tian", :added "0.8"}
+  (:require [clojure.string :as str]
+            [nrepl.middleware :as middleware :refer [set-descriptor!]]
+            [nrepl.middleware.session :as middleware.session]
+            [nrepl.misc :refer [response-for]]
+            [nrepl.transport :as t]))
+
+(defn require-resolve
+  [sym-str]
+  (let [sym (symbol sym-str)
+        ns (namespace sym)]
+    (when-not (find-ns sym)
+      (require (symbol ns))
+      (resolve sym))))
+
+(def ^:dynamic *state* nil)
+
+(defn wrap-ping
+  [h]
+  (fn [{:keys [op transport] :as msg}]
+    (case op
+      "ping"
+      (t/send transport (response-for msg {:pong "pong"
+                                           :status :done}))
+      (h msg))))
+
+(set-descriptor! #'wrap-ping
+                 {:require #{}
+                  :expects #{}
+                  :handles {"ping"
+                            {:doc "Ping"
+                             :require {}
+                             :returns {"status" "done"}}}})
+
+(defn unknown-op
+  "Sends an :unknown-op :error for the given message."
+  [{:keys [op transport] :as msg}]
+  (t/send transport (response-for msg :status #{:error :unknown-op :done} :op op)))
+
+(defn- update-stack!
+  [session middlewares]
+  (let [ctxcl  (.getContextClassLoader (Thread/currentThread))
+        alt-cl (when-let [classloader (:classloader (meta session))]
+                 (classloader))
+        cl     (or alt-cl
+                   ctxcl)]
+    (.setContextClassLoader (Thread/currentThread) cl)
+    (try
+      (with-bindings {clojure.lang.Compiler/LOADER cl}
+        (let [stack (->> middlewares
+                         (map (fn [middleware]
+                                (require-resolve
+                                 (if (var? middleware)
+                                   (str/replace (str middleware) "#'" "")
+                                   (str/replace middleware "#'" "")))))
+                         middleware/linearize-middleware-stack)]
+          (reset! *state* {:handler ((apply comp (reverse stack)) unknown-op)
+                           :stack   stack})))
+
+      (catch Throwable t
+        {:error        t
+         :alt-cl?      (boolean alt-cl)
+         :before-stack (map str (:stack @*state*))})
+      (finally
+        (.setContextClassLoader (Thread/currentThread) ctxcl)))))
+
+(defn wrap-dynamic-loader
+  [h]
+  (fn [{:keys [op transport session middlewares] :as msg}]
+    (case op
+      "add-middlewares"
```

Comment:
The plural of `middleware` is `middleware`. :-) https://english.stackexchange.com/questions/257120/middleware-vs-middlewares

Reply:
Will change :)

Reply:
Oh dear. Happy to change in new code, but "middlewares" is used quite liberally in the codebase + docs already.

In particular, `nrepl.server/default-middlewares` is an important part of the public API.

So it might be worth conceding to `middlewares` for nREPL purposes?

Reply:
I'll think about this. We've been inconsistent with the wording indeed, but I generally prefer to stick to the rules of natural languages. We can always have an alias for `default-middlewares` if it's our only public API transgression.

Reply:
I'd be supportive of that. I'll stick the the plural "middleware" in this PR, and we can switch everything else in a separate one for sake of commit history?

Reply:
Done